tests/unit/test_charm.py

```diff
@@ -334,3 +334,11 @@ def test_given_certificate_is_stored_when_on_certificates_relation_broken_then_c
 
         with pytest.raises(SecretNotFoundError):
             self.harness._backend.secret_get(label="certificate")
+
+    def test_given_csr_is_not_stored_when__request_certificate_called_then_runtime_error_is_not_raised(  # noqa: E501
+        self,
+    ):
+        with self.assertRaises(SecretNotFoundError):
```

Comment:
I don't understand this test.
Is an error raised?
why are we trying to get the secret from the secret backend?
The code seems to handle the case when the secret is not stored.
```
def _secret_exists(self, label: str) -> bool:
        """Returns whether a given secret exists.

        Args:
            label: Juju secret label

        Returns:
            bool: Whether the secret exists
        """
        try:
            self.model.get_secret(label=label)
            return True
        except SecretNotFoundError:
            return False
```

Reply:
With our change, we are not raising a  RuntimeError  is  if not self._csr_is_stored. This means model does not have a secret with CSR_SECRET_LABEL. I used the secret backend directly for validation which does not seems wrong to me. 
I think you are suggesting me to use alternative way for validation:
 self.assertEqual(self.harness.charm._secret_exists(label="csr"), False)  
 Is that right ?

Reply:
> Is an error raised?

If the secret is not there, SecretNotFoundError raises while checking secret backend.

Reply:
I just don't see the value in checking that calling `self.harness._backend.secret_get` directly raises an error or not.
That behaviour should be tested on Juju side.
And the way we use `get_secret` in our code, is that we check if throws an error or not, and if it does we catch it and return false.
I also don't recommend testing the private function `_secret_exists`.
`_request_certificate` is private too.
We should be testing it's side effects on the events we observe which are `config_changed` and `certificates_relation_joined`

Reply:
For example do we have a test like:
`test_given_no_csr_stored_when_config_changed_then_<whatever expected behaviour>`